### PR TITLE
Fix compilation on MinGW with pthread

### PR DESCRIPTION
**Pull request by orgads**

### CODE REVIEW

**Comment by BillyDonahue** [comment]
Can you describe the problem this is solving?
What is it about MinGW's pthread implementation that's causing trouble?

**Comment by orgads** [comment]
Done in the commit message.

**Comment by DanielSWolf** [comment]
:+1: I'm having the same build problem. This pull request fixes it for me.

**Comment by fredZen** [comment]
It fixes issue #606 

I hope the following excerpt from gtest-port.h clarifies the situation

    # if GTEST_HAS_NOTIFICATION_
    ...
    # elif GTEST_HAS_PTHREAD // line 1461
    // This branch is taken on MinGW, which prevents the next elif from getting taken
    ...
    # elif GTEST_OS_WINDOWS && !GTEST_OS_WINDOWS_PHONE && !GTEST_OS_WINDOWS_RT
    // On MinGW, this branch gets skipped, so AutoHandle doesn't get defined
    ...
    class GTEST_API_ AutoHandle {
    ...
    };
    ...
    # endif  // GTEST_HAS_NOTIFICATION_
    
    # if GTEST_HAS_MUTEX_AND_THREAD_LOCAL_
    ...
    # elif GTEST_OS_WINDOWS && !GTEST_OS_WINDOWS_PHONE && !GTEST_OS_WINDOWS_RT
    // This branch is taken even on MinGW
    ...
    class GTEST_API_ ThreadWithParamBase {
    ...
     private:
      AutoHandle thread_; // line 1782, breaks because AutoHandle does not name a type
    };

**Comment by fredZen** [comment]
See also #608

**Comment by BillyDonahue** [comment]
If MinGW has a broken Pthread, let's have GTEST_HAS_PTHREAD be unset for that platform, instead of having a compound conditional whenever we check for Pthread support. Would that work?

**Comment by orgads** [comment]
@BillyDonahue: Better now?

**Comment by BillyDonahue** [comment]
Don't undef and redefine PTHREAD macro. Just fold MINGW into the os test when we set it originally.

**Comment by orgads** [comment]
And what if the user chooses to define PTHREAD?

**Comment by BillyDonahue** [comment]
i mean GTEST_HAS_PTHREAD

**Comment by orgads** [comment]
Me too :)

Should it be used if the user explicitly defines it or not?

If not, how do you suggest to do it?

**Comment by orgads** [comment]
Let's start with being consistent within this header. Improvements can come later.

**Comment by orgads** [comment]
ping?

File: googletest/include/gtest/internal/gtest-port.h [hunk 1]
```diff
@@ -604,6 +604,14 @@ struct _RTL_CRITICAL_SECTION;
     || GTEST_OS_QNX || GTEST_OS_FREEBSD || GTEST_OS_NACL)
 #endif  // GTEST_HAS_PTHREAD
 
+// On MinGW, we can have both GTEST_OS_WINDOWS and GTEST_HAS_PTHREAD
+// defined, but we don't want to use MinGW's pthreads implementation, which
+// has conformance problems with some versions of the POSIX standard.
+#if GTEST_HAS_PTHREAD && GTEST_OS_WINDOWS_MINGW
```
Comment (BillyDonahue) [inline googletest/include/gtest/internal/gtest-port.h]: I don't want to undef something that was possibly set explicitly.
We should just deduce it correctly the first time.

```
#if !defined(GTEST_HAS_PTHREAD) && GTEST_OS_WINDOWS_MINGW
# define GTEST_HAS_PTHREAD 0
#endif
```

**Comment by KindDragon** [comment]
Duplicate of PR  #721

**Pull request closed without merging**